### PR TITLE
Conscrypt per factory [RFC]

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: samples/guide/src/main/java/okhttp3/recipes/CustomTrust.java [hunk 1]
```diff
@@ -44,9 +44,7 @@ public CustomTrust() {
     SSLSocketFactory sslSocketFactory;
     try {
       trustManager = trustManagerForCertificates(trustedCertificatesInputStream());
-      SSLContext sslContext = SSLContext.getInstance("TLS");
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/CustomTrust.java]: Can’t call internal APIs in examples

File: samples/guide/src/main/java/okhttp3/recipes/CustomCipherSuites.java [hunk 2]
```diff
@@ -55,7 +53,7 @@ public CustomCipherSuites() throws GeneralSecurityException {
         .build();
 
     X509TrustManager trustManager = defaultTrustManager();
-    SSLSocketFactory sslSocketFactory = defaultSslSocketFactory(trustManager);
+    SSLSocketFactory sslSocketFactory = Platform.get().getSslSocketFactory(trustManager);
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/CustomCipherSuites.java]: No internal APIs in examples

**Comment by yschimke** [comment]
I really like the client driven approach to passing a ConscryptSslSocketFactory to OkHttpClient.Builder, but in practice it is awkward because Platform is used as a JVM singleton and might have be Conscrypt, Jdk9, Platform or Jdk8Platform.

I think it only makes sense if we allow clients to select Platform per client instance.

**Comment by swankjesse** [comment]
Maybe we should have a different Platform for different security providers? Maybe a method that retrieves a platform from a given SSLSocketFactory?

**Comment by yschimke** [comment]
No short term fix here.  Can be worked around by apps using Conscrypt in two places, by setting one per connection and the other using global default.  For OkHttp this would mean supplying your own ssl socket factor that sets this.

**Pull request closed without merging**